Update export types

dbritto-dev opened the pull request:
## Related Issues or Discussions

Fixes #2163

## Summary

The types for should be under each type of import. For instance "import", "module", "require" and "default"

## Check List

- [x] `yarn run prettier` for formatting code and docs

dbritto-dev commented:
@dai-shi here you go

dai-shi commented:
> ## Related Issues or Discussions
> Fixes #2164

Do you mean #2163 ?

dai-shi commented on package.json:
```diff
@@ -21,12 +21,13 @@
   "exports": {
     "./package.json": "./package.json",
     ".": {
+      "types": "./esm/index.d.ts",
```
Do you need this? Also, I just noticed that we didn't change other than ".".

dbritto-dev replied:
For some reason esm types throw some errors, that set up the `index.d.ts` as default for everything but internally we use the ones that we had under types on each config

dbritto-dev commented:
> > ## Related Issues or Discussions
> 
> > Fixes #2164
> 
> 
> 
> Do you mean #2163 ?

Yeah, that one

dbritto-dev commented:
@dai-shi all yours

dai-shi commented on package.json:
```diff
@@ -26,85 +26,27 @@
-    },
-    "./traditional": {
-      "types": "./traditional.d.ts",
+    "./*": {
```
This requires newer version of Node, and we will be doing this change in v5.

dai-shi replied:
@dbritto-dev We are waiting for the revert.

dbritto-dev replied:
@dai-shi done

dai-shi replied:
Ooops, sorry. I don't mean to revert #2157 fully.
Can you try what you think should work, and just avoid using `"./*"`.

dbritto-dev replied:
In that we case we should apply the same change for each inner package until we can use the `./*`

dai-shi replied:
Yes, that what I meant.

dbritto-dev replied:
Sure, let me do it

dai-shi commented on package.json:
```diff
@@ -26,85 +26,27 @@
       },
       "module": {
-        "types": "./index.d.ts",
+        "types": "./esm/index.d.ts",
```
Is this change the key to fix the issue? I mean the file content must be the same, so the file path was the root cause, right? Why wasn't it an issue before #2157?

It feels like reverting #2157 should be appropriate, if we **don't know** why it caused such an issue.

dbritto-dev replied:
@dai-shi btw, I updated the exports to match jotai's and valtio's exports

dbritto-dev replied:
@dai-shi but we can revert the changes

dai-shi replied:
Yeah > https://github.com/pmndrs/zustand/pull/2170#discussion_r1379421218

dai-shi commented:
btw, @dbritto-dev do you understand why #2157 caused such an issue? It seems like TS picks different files for index and middleware, but I'm not sure how that could happen. It sounds like TS ignore "types" field, doesn't it?

dbritto-dev commented:
> btw, @dbritto-dev do you understand why #2157 caused such an issue? It seems like TS picks different files for index and middleware, but I'm not sure how that could happen. It sounds like TS ignore "types" field, doesn't it?

@dai-shi I understand the issue in the case of inner packages when you use "types" outside of "import", "module", "require" or "default" ts use by default that types ignoring where you are importing the packages when we added the new changes for the main package we fixed that but the inner packages still pointing to "d.ts" even if you are importing them from "esm" that's the conflict

dbritto-dev commented:
@dai-shi all yours

dai-shi commented:
> in the case of inner packages when you use "types" outside of "import", "module", "require" or "default" ts use by default that types ignoring where you are importing the packages;

So, the "types" line right under "exports" in v4.4.4 package.json was simply ignored? That explains it.

dbritto-dev commented:
@dai-shi would you mind taking care of these changes? I tested locally and works well

dai-shi commented:
Yes, it looks good.
I hope to get one or two more people to try it. https://github.com/pmndrs/zustand/discussions/2163#discussioncomment-7457555

dai-shi reviewed (approved):
The changes look good.

LarsArtmann commented:
Thank you for fixing my build :) 👍

sanduluca commented:
Was this released as I still have the issue with immer and `"zustand": "^5.0.3"`

![image](https://github.com/user-attachments/assets/aa6edb9a-68ae-4d9c-9c6b-81a659a8149b)

dbritto-dev commented:
> Was this released as I still have the issue with immer and `"zustand": "^5.0.3"`
> 
> ![image](https://private-user-images.githubusercontent.com/56228534/406765439-aa6edb9a-68ae-4d9c-9c6b-81a659a8149b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc5MzA5MTksIm5iZiI6MTczNzkzMDYxOSwicGF0aCI6Ii81NjIyODUzNC80MDY3NjU0MzktYWE2ZWRiOWEtNjhhZS00ZDljLTljNmItODFhNjU5YTgxNDliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI2VDIyMzAxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU4ZWYwYWY0NzhlODYyZTdhMWNjMWU3NDZiODI4OTg4YTA4MWZhYmJmNWVmZjQ4NjlmOGE5NzRjZTE4N2Q3ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MDzOC9uuODCfC6_XMyQFkxTK9oun46NEWFLDOP0u7ig)

@sanduluca that doesn't look right to me, `addSound` method should be implemented keeping in mind `immer`

sanduluca commented:
In the image i was showing the that the immer is generating the error (before refactoring). 

But the point is still that it shows type errors with immer even with the example from docs.

![image](https://github.com/user-attachments/assets/76fa50da-10ae-4b42-97aa-2a1a60f1db3b)

<details><summary>Code and error message</summary>
<p>

```ts
type State = {
  count: number;
};

type Actions = {
  increment: (qty: number) => void;
  decrement: (qty: number) => void;
};

export const useCountStore = create<State & Actions>()(
  immer((set) => ({
    count: 0,
    increment: (qty: number) =>
      set((state) => {
        state.count += qty;
      }),
    decrement: (qty: number) =>
      set((state) => {
        state.count -= qty;
      }),
  }))
);
```
```bash
Argument of type 'StateCreator<State & Actions, [], [["zustand/immer", never]]>' is not assignable to parameter of type 'StateCreator<State & Actions, [], [never, unknown][]>'.
  Type 'StateCreator<State & Actions, [], [["zustand/immer", never]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'.
    Types of property '$$storeMutators' are incompatible.
      Type '[["zustand/immer", never]] | undefined' is not assignable to type '[never, unknown][] | undefined'.
        Type '[["zustand/immer", never]]' is not assignable to type '[never, unknown][]'.
          Type '["zustand/immer", never]' is not assignable to type '[never, unknown]'.
            Type at position 0 in source is not compatible with type at position 0 in target.
              Type '"zustand/immer"' is not assignable to type 'never'.
```

![image](https://github.com/user-attachments/assets/aabda4ea-c860-4db1-a530-1a60c60c76e8)

</p>
</details>

dbritto-dev commented:
> In the image i was showing the that the immer is generating the error (before refactoring). 
> 
> 
> 
> But the point is still that it shows type errors with immer even with the example from docs.
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/76fa50da-10ae-4b42-97aa-2a1a60f1db3b)
> 
> 
> 
> <details><summary>Code, error message, package.json</summary>
> 
> <p>
> 
> 
> 
> ```ts
> 
> type State = {
> 
>   count: number;
> 
> };
> 
> 
> 
> type Actions = {
> 
>   increment: (qty: number) => void;
> 
>   decrement: (qty: number) => void;
> 
> };
> 
> 
> 
> export const useCountStore = create<State & Actions>()(
> 
>   immer((set) => ({
> 
>     count: 0,
> 
>     increment: (qty: number) =>
> 
>       set((state) => {
> 
>         state.count += qty;
> 
>       }),
> 
>     decrement: (qty: number) =>
> 
>       set((state) => {
> 
>         state.count -= qty;
> 
>       }),
> 
>   }))
> 
> );
> 
> ```
> 
> ```bash
> 
> Argument of type 'StateCreator<State & Actions, [], [["zustand/immer", never]]>' is not assignable to parameter of type 'StateCreator<State & Actions, [], [never, unknown][]>'.
> 
>   Type 'StateCreator<State & Actions, [], [["zustand/immer", never]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'.
> 
>     Types of property '$$storeMutators' are incompatible.
> 
>       Type '[["zustand/immer", never]] | undefined' is not assignable to type '[never, unknown][] | undefined'.
> 
>         Type '[["zustand/immer", never]]' is not assignable to type '[never, unknown][]'.
> 
>           Type '["zustand/immer", never]' is not assignable to type '[never, unknown]'.
> 
>             Type at position 0 in source is not compatible with type at position 0 in target.
> 
>               Type '"zustand/immer"' is not assignable to type 'never'.
> 
> ```
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/aabda4ea-c860-4db1-a530-1a60c60c76e8)
> 
> 
> 
> </p>
> 
> </details> 

@sanduluca would you mind opening a discussion?

Ray5055 commented:
Hi! Is there any updates on this matter? I'm facing similar issues however mine is with persist rather than immer

```
Argument of type 'StateCreator<AssetState, [], [["zustand/persist", AssetState]]>' is not assignable to parameter of type 'StateCreator<AssetState, [], [never, unknown][]>'.
  Type 'StateCreator<AssetState, [], [["zustand/persist", AssetState]]>' is not assignable to type '{ $$storeMutators?: [never, unknown][] | undefined; }'.
    Types of property '$$storeMutators' are incompatible.
      Type '[["zustand/persist", AssetState]] | undefined' is not assignable to type '[never, unknown][] | undefined'.
        Type '[["zustand/persist", AssetState]]' is not assignable to type '[never, unknown][]'.
          Type '["zustand/persist", AssetState]' is not assignable to type '[never, unknown]'.
            Type at position 0 in source is not compatible with type at position 0 in target.
              Type '"zustand/persist"' is not assignable to type 'never'.ts(2345)
```
